### PR TITLE
Postgres events

### PR DESCRIPTION
**Pull request by espadolini**
Subordinate to the RFD at #28355; opening for review as I'm expecting only minor UX/config changes at this point.

This PR adds an audit log implementation that uses PostgreSQL as storage. In addition to using fixed credentials, the implementation can also fetch Azure AD tokens, to log in to Azure Database for PostgreSQL database clusters.

### CODE REVIEW

File: lib/events/pgevents/pgevents.go
Comment (Tener) [inline lib/events/pgevents/pgevents.go:220]: `periodicCtx` may be unused with disabled cleanup, but we should clean it up anyway, so it shouldn't leak or be an actual drain on resources.

File: lib/events/pgevents/pgevents.go
Comment (Tener) [inline lib/events/pgevents/pgevents.go:341]:
```suggestion
// periodicCleanup removes events past the retention period from the table,
```

`their` suggests that the retention period is per item, which is not the case.

File: lib/events/pgevents/utils.go
Comment (Tener) [inline lib/events/pgevents/utils.go:53]: What do you think about adding some tests here? A fuzzing test would be nice as well, e.g. to test that `toNextKey` and `fromStartKey` can be combined to form (partial) identity?

File: lib/events/pgevents/common/azure.go
Comment (espadolini) [reply lib/events/pgevents/common/azure.go:74]: It's possible at startup, but I don't know if it's a problem in practice; the Azure IDMS rate limits claim that a burst of 20 requests per second is fine. 🤷 

What structure were you thinking of? The `FnCache` wants a fixed TTL, whereas our TTL is decided by the token that we get.

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:220]: If we only create the context when the cleanup is not disabled we must check that `cancel` is non-nil in `Close`; I found it cleaner to just always have it.

File: lib/events/pgevents/common/azure.go
Comment (espadolini) [reply lib/events/pgevents/common/azure.go:74]: I'm also just going to fetch the token once at configuration time to avoid the initial burst of fetches. 😌

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:307]: Where are we assuming that? We run through the schemas if the version we get is lesser than the maximum version in the `schemas` slice.

File: lib/events/pgevents/common/azure.go
Comment (espadolini) [reply lib/events/pgevents/common/azure.go]: I remember having checked about a year ago and I don't remember it being like this, but `azidentity` token credentials actually do the caching and the singleflighting for us, so I simplified the whole thing to just call `GetToken` every time in fc1717bfa1.

File: lib/events/pgevents/common/azure.go
Comment (espadolini) [reply lib/events/pgevents/common/azure.go]: Greatly simplified in fc1717bfa1.

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:313]: Added in 6d5adccf82

File: lib/events/pgevents/utils.go
Comment (espadolini) [reply lib/events/pgevents/utils.go:53]: I added a roundtrip test that tests uniformly random times and UUIDs, and pagination keys in f5e37bc6d4. WDYT?

Comment (Tener) [reply lib/events/pgevents/utils.go:53]: Looks good.

**Review by Tener — APPROVED** [review]

File: go.mod
Comment (jakule) [inline go.mod:105]: Do we need both packages? Shouldn't we stick to v4 for now?

File: lib/events/pgevents/common/azure.go
Comment (jakule) [inline lib/events/pgevents/common/azure.go:29]: Delete the whole file? Can you create an issue for that and assign it to someone who should be responsible for that?

File: lib/events/pgevents/common/utils.go
Comment (jakule) [inline lib/events/pgevents/common/utils.go:33]: The same as above. Also looks like all that code will be removed "soon" after it's merged?

File: lib/events/pgevents/common/utils.go
Comment (jakule) [inline lib/events/pgevents/common/utils.go:39]: nit: Is there a constant for that?

File: lib/events/pgevents/common/utils.go
Comment (jakule) [inline lib/events/pgevents/common/utils.go:72]: What if I change the name in the configuration after it's already been created?

File: lib/events/pgevents/common/utils.go
Comment (jakule) [inline lib/events/pgevents/common/utils.go:79]:
```suggestion
			log.WithError(err).Debug("Error creating database due to insufficient privileges.")
```

File: lib/events/pgevents/pgevents.go
Comment (jakule) [inline lib/events/pgevents/pgevents.go:49]: Casually also called a year :sweat_smile:

File: lib/events/pgevents/pgevents.go
Comment (jakule) [inline lib/events/pgevents/pgevents.go:303]: From what I understand, we will scale auth to 1, run this and scale back? That also means that rollback after won't be possible (when the cheam changes)?

File: lib/events/pgevents/pgevents.go
Comment (jakule) [inline lib/events/pgevents/pgevents.go:341]: Shouldn't you commit after the migration? Otherwise, other instances won't be able to see the changes.

File: lib/events/pgevents/pgevents.go
Comment (jakule) [inline lib/events/pgevents/pgevents.go:464]: Doing string operation instead of using a parametrized question won't trash the internal DB cache?

**Review by jakule — APPROVED** [review]
Approved with a few comments to not block this PR (overall looks good).

File: lib/events/pgevents/common/utils.go
Comment (espadolini) [reply lib/events/pgevents/common/utils.go:33]: Either #29025 gets merged first, in which case I'll update this PR to use `lib/backend/pgbk/common`, or this PR gets merged first, in which case I'll update #29025 to delete `lib/events/pgevents/common` in favor of `lib/backend/pgbk/common`. A better name/place to put that is very welcome, tho 😅

File: go.mod
Comment (espadolini) [reply go.mod:105]: Why?

File: lib/events/pgevents/common/azure.go
Comment (espadolini) [reply lib/events/pgevents/common/azure.go:29]: replied in https://github.com/gravitational/teleport/pull/29057#discussion_r1265803379

File: lib/events/pgevents/common/utils.go
Comment (espadolini) [reply lib/events/pgevents/common/utils.go:72]: There's no way to change the configuration without restarting the audit log (or Teleport as a whole, really).

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:303]: I'm not really expecting us to do changes to the schema, but yes, a low-level schema change will not be rollbackable without some manual intervention on the database. It should be fine to start multiple new auths all at the same version however, the setup transaction will just be retried I reckon.

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:464]: It's not that bad now that there's only 16 possibilities (a lot of which, I bet, are actually never exercised), but I'll leave a comment to run the query in `pgx.QueryExecModeExec` if we eventually decide to do server-side filtering with arbitrary `WhereCond`s.

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:49]: updated in 445c9d7f8d 🤔

File: lib/events/pgevents/common/utils.go
Comment (espadolini) [reply lib/events/pgevents/common/utils.go:39]: I can't find a good name for it, even the postgres docs don't give it a name other than "a database named `postgres`, which is meant as a default database for use by utilities, users and third party applications".

File: lib/events/pgevents/pgevents.go
Comment (espadolini) [reply lib/events/pgevents/pgevents.go:464]: 3d7521ac7889c8e457c96bb18ab7e948f270c62c

File: go.mod
Comment (jakule) [reply go.mod:105]: Because I'm pretty sure that no one tests if v4 and v5 are compatible, and using both may have some unexpected side effects?

Comment (espadolini) [reply go.mod:105]: They don't have to be compatible, they're completely independent implementations - that's the whole point of go module majors, you get to make breaking changes because the two versions can be used at the same time.

The only potential source of weirdness is if both `jackc/pgx/v4/stdlib` and `jackc/pgx/v5/stdlib` are imported, as that would make it unclear ([until Go 1.21](https://tip.golang.org/doc/go1.21#language)) which of the two `database/sql` drivers gets registered as `"pgx“` (they both also register `"pgx/v4"` and `"pgx/v5"` respectively). We use neither in the code, the only use of `github.com/jackc/pgx/v4/stdlib` is in a code example in the docs.

`pgx/v5` has a couple of new features that I'm relying on, `pgx.NamedArgs` being the harder to replace in a confidently safe way, and improved Batch ergonomics being another (not used here but in pgbk). I'd very much prefer to not have to update the code to go back to a previous major version.

File: lib/events/pgevents/common/utils.go
Comment (espadolini) [reply lib/events/pgevents/common/utils.go:33]: Moved it to `lib/backend/pgbk/common` now, it makes it easier to keep it in sync between the two PRs anyway.

**Review by jakule — APPROVED** [review]

File: lib/events/pgevents/utils_test.go
Comment (espadolini) [reply lib/events/pgevents/utils_test.go:28]: 15c2ed9990e9e796324e05c0a67bc7fdc5ce4db4

**Pull request merged**
